### DISCUSSION TITLE
What do you think about this idea?

### DISCUSSION THREAD

**Discussion by frederikhors** [paoloricciuti/sveltekit-search-params#84] [Q&A]
Can I ask you what do you think about https://github.com/sveltejs/kit/discussions/12392?

**Comment by paoloricciuti** [comment]
Well this package kinda does what you are trying to do. Is your need svelte 5? I'm kinda working on a svelte 5 version.

Btw I think a safer bet than calling `goto` on an effect would be to derive from it and call it when the user set the value rather then setting the value and then redirect.

**Reply by frederikhors** [reply]
Yeah, my need is Svelte 5. But also:

1. to assign a default value to be used with `$props()`:

    ```ts
    let {
        pagination = params.pagination,
        condition = params.condition,
        searchText = params.searcText
    } = $props();
    ```

1. to avoid the call to `$effect(() => {})` to sync objects with URL:

    ```ts
    $effect(() => {
        if (!!condition) {
            const url = new URL(window.location.toString());

            const paramCondition = url.searchParams.get('condition');

            const actualCondition = JSON.stringify(condition);

            if (paramCondition !== actualCondition) {
                url.searchParams.set('condition', JSON.stringify(condition));

                goto(url, { replaceState: false, noScroll: true });
            }
        }
    });
    ```

1. to avoid the call to `afterNavigate()` for the same reason.

Do you think is this possible?

> Btw I think a safer bet than calling goto on an effect would be to derive from it and call it when the user set the value rather then setting the value and then redirect.

Can you write a small example here for this? I don't understand what you mean.

I'm ready to try whatever you want IMMEDIATELY.

Because I need it ASAP.

Thank you very much!

**Reply by paoloricciuti** [reply]
Well it will probably take more the quick code I can write now on my phone but basically what you can do is something like this
```ts
function syncParams(params){
    const fromUrl = $derived(decodeParams($page.url.searchParams));
    const retval = {};
    for( let key in params){
        Object.defineProperty(retval, key, {
            get(){ return fromUrl[key]; }
            set(newVal){
                goto(calculateNewUrl(newVal));
            },
            enumerable: true,
        });
    }
    return retval;
}
```

**Reply by frederikhors** [reply]
Wow. This is beautiful.

But I need to work with binding, something like:

```svelte
let {
    condition = params.condition,
} = $props();

<input type="text" bind:value={condition.name} />
```

Do you think that can work?

**Reply by paoloricciuti** [reply]
You probably need to first pass condition to the function and then bind to the return value of the function

**Reply by frederikhors** [reply]
I don't know if I understood correctly what you're saying.

In any case, I don't want to take up your time now that you're not at the PC.

When do you think you can write something that I can then try right after?

Among other things, I think we can use https://www.sveltelab.dev instead of Stackblitz because it is much faster and above all it shows the URL perfectly which is what we need most.

**Reply by frederikhors** [reply]
@paoloricciuti I created https://www.sveltelab.dev/84hoezlz83nben0.

Under the route `/paolo` you can find the `+page.svelte` file with your `syncParams` code. I can't make it work tough. :(

**Reply by frederikhors** [reply]
@paoloricciuti, do you think you can find a few minutes to solve this with your mastery?

It would benefit many people who have the same problem with SvelteKit.

**Comment by paoloricciuti** [comment]
Hey i just put up a PR for the svelte 5 version of the library if you want to take a look!

https://github.com/paoloricciuti/sveltekit-search-params/pull/126

You can install the package like this

```
pnpm add https://pkg.pr.new/sveltekit-search-params@126
```
